lib/files.js

```diff
@@ -52,20 +64,26 @@ var putJSON = function(info, callback) {
     if (!doc.name || doc.error) {
         return callback(doc.error);
     }
-    var file = path.join(options.dir, doc.name, 'index.json');
-    log.info('[' + info.seq + '/' + info.latestSeq + ']', 'writing json for', doc.name, 'to', file);
-    process.nextTick(function() {
-        mkdirp(path.dirname(file), function() {
-            process.nextTick(function() {
-                fs.writeFile(file, JSON.stringify(doc, null, 4) + '\n', function(err) {
-                    if (err) {
-                        return putJSON(info, callback);
-                    }
-                    if (!info.versions || !info.versions.length) {
-                        return callback();
-                    }
-                    process.nextTick(function() {
-                        async.eachLimit(info.versions, 5, putPart, callback);
+    var putAllParts = function(err) {
+        if (err) {
+            return callback(err);
+        }
+        async.eachLimit(info.versions, 5, putPart, callback);
+    };
+    hooks.indexJson(info, putAllParts, function() {
+        var file = path.join(options.dir, doc.name, 'index.json');
+        log.info('[' + info.seq + '/' + info.latestSeq + ']', 'writing json for', doc.name, 'to', file);
```

Comment:
Since this `info` is by reference, maybe we should put `info.seq` and `info.latestSeq` into private vars here so they can't accidentally move the sequence that we are tracking. Changing those numbers would totally change the area that it's following and could really screw things up.